### PR TITLE
raftstore: speedup compact

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @siddontang PTAL

### CODE REVIEW

**Comment by siddontang** [comment]
LGTM

File: src/raftstore/store/keys.rs [hunk 1]
```diff
@@ -90,6 +90,16 @@ pub fn raft_log_key(region_id: u64, log_index: u64) -> Vec<u8> {
     key
 }
 
+/// Get the log index from raft log key generated by `raft_log_key`.
+pub fn raft_log_index(key: &[u8]) -> Result<u64> {
+    let exp_key_len = REGION_RAFT_PREFIX_KEY.len() + mem::size_of::<u64>() + mem::size_of::<u8>() +
```
Comment (ngaut) [inline src/raftstore/store/keys.rs]: What does exp mean ?

Comment (siddontang) [inline src/raftstore/store/keys.rs]: I think is expect

Comment (ngaut) [inline src/raftstore/store/keys.rs]: We should rename to expect, it makes me confused.

File: src/raftstore/store/worker/compact.rs [hunk 2]
```diff
@@ -46,34 +46,52 @@ impl Display for Task {
...
+        let start_key = keys::raft_log_key(task.region_id, 0);
+        let mut first_idx = task.compact_idx;
+        if let Some((k, _)) = box_try!(task.engine.seek(&start_key)) {
+            first_idx = box_try!(keys::raft_log_index(&k));
+        }
+        if first_idx >= task.compact_idx {
+            info!("no need to compact");
+            return Ok(0);
+        }
+        let w = WriteBatch::new();
```
Comment (ngaut) [inline src/raftstore/store/worker/compact.rs]: s/w/wb/

File: src/raftstore/store/worker/compact.rs [hunk 3]
```diff
@@ -46,34 +46,52 @@ impl Display for Task {
...
+            display("compact failed {:?}", err)
+        }
+    }
+}
+
 pub struct Runner;
 
-impl Runner {}
+impl Runner {
+    fn compact(&mut self, task: Task) -> Result<u64, Error> {
```
Comment (ngaut) [inline src/raftstore/store/worker/compact.rs]: Please add comment for return result.

**Comment by ngaut** [comment]
LGTM

**Pull request merged**